### PR TITLE
Support for product_type = degree in Django ORM POC for SFMC Catalog

### PR DESCRIPTION
**Pull request by hamza-56**
### [PROD-4153](https://2u-internal.atlassian.net/browse/PROD-4153)

This PR enhances the existing management command for populating the Course Catalog by adding support for the Degree Catalog.

Key Updates:

- Degree Catalog Support: Added logic to populate required fields from the Program Model into a CSV file.
- Filtering: Introduced filters for `product_type` and `product_source`.
- Google Sheets Update: Integrated with gspread to update an existing Google Sheet.
- Unit Tests: Added tests to ensure functionality for the Degree Catalog.

### CODE REVIEW

File: course_discovery/apps/course_metadata/management/commands/populate_product_catalog.py [hunk 1]
```diff
@@ -163,7 +205,7 @@ def handle(self, *args, **options):
                     output_writer = self.write_csv_header(output_file)
                     for product in products:
                         try:
-                            output_writer.writerow(self.get_transformed_data(product))
+                            output_writer.writerow(self.get_transformed_data(product, product_type))
                         except Exception as e:  # pylint: disable=broad-exception-caught
```
Comment (AfaqShuaib09) [inline course_discovery/apps/course_metadata/management/commands/populate_product_catalog.py:200]: Update the get_transformed_data parameter in the use_gspread_client scenario. This will fix the failed test case. You haven't updated the function call here
https://github.com/openedx/course-discovery/blob/master/course_discovery/apps/course_metadata/management/commands/populate_product_catalog.py#L174

File: course_discovery/apps/course_metadata/management/commands/populate_product_catalog.py [hunk 2]
```diff
@@ -112,28 +132,49 @@ def write_csv_header(self, output_csv):
...
+                "Organizations Abbr": ", ".join(org.key for org in authoring_orgs),
+                "Languages": ", ".join(language.code for language in product.languages),
+                "Subjects": ", ".join(subject.name for subject in product.subjects),
+                "Subjects Spanish": ", ".join(
+                    translation.name for subject in product.subjects
+                    for translation in subject.spanish_translations
+                ),
+                "Marketing URL": product.marketing_url,
+                "Marketing Image": product.card_image.url if product.card_image else "",
+            }
```
Comment (DawoudSheraz) [inline course_discovery/apps/course_metadata/management/commands/populate_product_catalog.py]: Most of the fields are the same across both product types. Only some differ. It is better to have a base dict and update that base dict with only those fields that differ across the product types

File: course_discovery/apps/course_metadata/management/commands/populate_product_catalog.py [hunk 3]
```diff
@@ -100,6 +100,26 @@ def get_products(self, product_type, product_source):
                 'subjects',
                 subject_translations
             )
+        elif product_type == 'degree':
+            queryset = Program.objects.marketable() \
+                .exclude(degree__isnull=True) \
+                .select_related('partner', 'type')
```
Comment (AfaqShuaib09) [inline course_discovery/apps/course_metadata/management/commands/populate_product_catalog.py]: nit: No need to break line here, line length is already less than 120 characters

Comment (hamza-56) [reply course_discovery/apps/course_metadata/management/commands/populate_product_catalog.py]: The line breaks were added to enhance readability.

File: course_discovery/apps/course_metadata/management/commands/populate_product_catalog.py [hunk 4]
```diff
@@ -112,28 +132,43 @@ def write_csv_header(self, output_csv):
...
 
-    def get_transformed_data(self, product):
+    def get_transformed_data(self, product, product_type):
         """
         Transforms the product data for product's catalog
         """
         authoring_orgs = product.authoring_organizations.all()
-        return {
+
+        common_data = {
```
Comment (DawoudSheraz) [inline course_discovery/apps/course_metadata/management/commands/populate_product_catalog.py]: nit: data should be enough here. The code structure is good enough to explain the base data and specific data against product_type.

File: course_discovery/apps/course_metadata/management/commands/tests/test_populate_product_catalog.py [hunk 5]
```diff
@@ -65,6 +74,33 @@ def test_populate_product_catalog(self):
                 self.assertIn("Marketing URL", row)
                 self.assertIn("Marketing Image", row)
 
+    def test_populate_product_catalog_for_degree(self):
+        """
+        Test populate_product_catalog command for degree product type and verify data has been populated successfully
+        """
+        with NamedTemporaryFile(mode="w", delete=False) as output_csv:
```
Comment (DawoudSheraz) [inline course_discovery/apps/course_metadata/management/commands/tests/test_populate_product_catalog.py]: if delete=False here, at what point does that temporary file delete?

File: course_discovery/apps/course_metadata/management/commands/tests/test_populate_product_catalog.py [hunk 6]
```diff
@@ -140,6 +176,32 @@ def test_get_transformed_data(self):
             "Marketing Image": (product.image.url if product.image else ""),
         }
 
+    def test_get_transformed_data_for_degree(self):
+        """
+        Verify get_transformed_data method is working correctly for degree
+        """
+        product = self.degrees[0]
+        command = Command()
```
Comment (DawoudSheraz) [inline course_discovery/apps/course_metadata/management/commands/tests/test_populate_product_catalog.py:276]: why is Command() created here?

File: course_discovery/apps/course_metadata/management/commands/tests/test_populate_product_catalog.py [hunk 7]
```diff
@@ -65,6 +74,33 @@ def test_populate_product_catalog(self):
...
+                self.assertIn("UUID", row)
+                self.assertIn("Title", row)
+                self.assertIn("Organizations Name", row)
+                self.assertIn("Organizations Logo", row)
+                self.assertIn("Organizations Abbr", row)
+                self.assertIn("Languages", row)
+                self.assertIn("Subjects", row)
+                self.assertIn("Subjects Spanish", row)
+                self.assertIn("Marketing URL", row)
+                self.assertIn("Marketing Image", row)
```
Comment (DawoudSheraz) [inline course_discovery/apps/course_metadata/management/commands/tests/test_populate_product_catalog.py]: Does this check the value too? Or only the existence of the key?

File: course_discovery/apps/course_metadata/management/commands/tests/test_populate_product_catalog.py [hunk 8]
```diff
@@ -65,6 +74,33 @@ def test_populate_product_catalog(self):
                 self.assertIn("Marketing URL", row)
                 self.assertIn("Marketing Image", row)
 
+    def test_populate_product_catalog_for_degree(self):
```
Comment (DawoudSheraz) [inline course_discovery/apps/course_metadata/management/commands/tests/test_populate_product_catalog.py]: This test should either have ddt or there can be 1-2 more tests to verify other conditions of the command (degree being null, marketable, etc.).

File: course_discovery/apps/course_metadata/management/commands/tests/test_populate_product_catalog.py (hunk 6 above)
Comment (hamza-56) [reply course_discovery/apps/course_metadata/management/commands/tests/test_populate_product_catalog.py:276]: we're using `command.get_transformed_data(product, "degree")` on line 186

Comment (DawoudSheraz) [reply course_discovery/apps/course_metadata/management/commands/tests/test_populate_product_catalog.py:276]: Right. How does it evaluate that the command to execute is populate catalog command?

**Review by DawoudSheraz — COMMENTED** [review]
Please update PR description to include the relevant context.

File: course_discovery/apps/course_metadata/management/commands/tests/test_populate_product_catalog.py (hunk 6 above)
Comment (hamza-56) [reply course_discovery/apps/course_metadata/management/commands/tests/test_populate_product_catalog.py:276]: `from course_discovery.apps.course_metadata.management.commands.populate_product_catalog import Command`

File: course_discovery/apps/course_metadata/management/commands/tests/test_populate_product_catalog.py (hunk 7 above)
Comment (hamza-56) [reply course_discovery/apps/course_metadata/management/commands/tests/test_populate_product_catalog.py]: Updated to check existence of key

File: course_discovery/apps/course_metadata/management/commands/tests/test_populate_product_catalog.py (hunk 8 above)
Comment (hamza-56) [reply course_discovery/apps/course_metadata/management/commands/tests/test_populate_product_catalog.py]: Added

File: course_discovery/apps/course_metadata/management/commands/tests/test_populate_product_catalog.py [hunk 9]
```diff
@@ -51,19 +60,137 @@ def test_populate_product_catalog(self):
...
+        - Degrees with status not set to ProgramStatus.Active
+        - Degrees with an empty marketing_slug
+        """
+        non_marketable_degrees = [
+            DegreeFactory.create(
+                product_source=self.source,
+                partner=self.partner,
+                additional_metadata=None,
+                type=self.program_type,
+                status=ProgramStatus.Unpublished,  # Unpublished status
```
Comment (DawoudSheraz) [inline course_discovery/apps/course_metadata/management/commands/tests/test_populate_product_catalog.py]: nit: the comment `# Unpublished status` is unnecessary.

**Review by DawoudSheraz — APPROVED** [review]

File: course_discovery/apps/course_metadata/management/commands/tests/test_populate_product_catalog.py [hunk 10]
```diff
@@ -51,19 +60,137 @@ def test_populate_product_catalog(self):
...
+                marketing_slug="valid-slug-1",
+                title="Non-Marketable Degree - Unpublished Status"
+            ),
+            DegreeFactory.create(
+                product_source=self.source,
+                partner=self.partner,
+                additional_metadata=None,
+                type=self.program_type,
+                status=ProgramStatus.Retired,  # Retired status
```
Comment (AfaqShuaib09) [inline course_discovery/apps/course_metadata/management/commands/tests/test_populate_product_catalog.py]: nit: remove status inline comments

File: course_discovery/apps/course_metadata/management/commands/tests/test_populate_product_catalog.py [hunk 11]
```diff
@@ -51,19 +60,137 @@ def test_populate_product_catalog(self):
...
+                marketing_slug="valid-slug-2",
+                title="Non-Marketable Degree - Retired Status"
+            ),
+            DegreeFactory.create(
+                product_source=self.source,
+                partner=self.partner,
+                additional_metadata=None,
+                type=self.program_type,
+                status=ProgramStatus.Deleted,  # Deleted status
```
Comment (AfaqShuaib09) [inline course_discovery/apps/course_metadata/management/commands/tests/test_populate_product_catalog.py]: same here

File: course_discovery/apps/course_metadata/management/commands/tests/test_populate_product_catalog.py [hunk 12]
```diff
@@ -51,19 +60,137 @@ def test_populate_product_catalog(self):
...
+                marketing_slug="valid-slug-3",
+                title="Non-Marketable Degree - Deleted Status"
+            ),
+            DegreeFactory.create(
+                product_source=self.source,
+                partner=self.partner,
+                additional_metadata=None,
+                type=self.program_type,
+                status=ProgramStatus.Active,  # Active, but empty marketing slug
```
Comment (AfaqShuaib09) [inline course_discovery/apps/course_metadata/management/commands/tests/test_populate_product_catalog.py]: same here

File: course_discovery/apps/course_metadata/management/commands/tests/test_populate_product_catalog.py [hunk 13]
```diff
@@ -51,19 +60,137 @@ def test_populate_product_catalog(self):
...
+                marketing_slug="",  # Empty marketing_slug
+                title="Non-Marketable Degree - Empty Slug"
+            )
+        ]
+
+        # Create a degree that should be included (marketable)
```
Comment (AfaqShuaib09) [inline course_discovery/apps/course_metadata/management/commands/tests/test_populate_product_catalog.py]: remove this inline comment, variable name is good enough

File: course_discovery/apps/course_metadata/management/commands/tests/test_populate_product_catalog.py [hunk 14]
```diff
@@ -51,19 +60,137 @@ def test_populate_product_catalog(self):
...
+        marketable_degree = DegreeFactory.create(
+            product_source=self.source,
+            partner=self.partner,
+            additional_metadata=None,
+            type=self.program_type,
+            status=ProgramStatus.Active,  # Active status
+            marketing_slug="valid-marketing-slug",  # Valid marketing_slug
```
Comment (AfaqShuaib09) [inline course_discovery/apps/course_metadata/management/commands/tests/test_populate_product_catalog.py]: nit: remove inline comment

**Review by AfaqShuaib09 — APPROVED** [review]
Few nits to address, otherwise LGTM
